internal/integrationtest/util.go

```diff
@@ -0,0 +1,255 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package integrationtest
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.uber.org/yarpc"
+	"go.uber.org/yarpc/api/peer"
+	"go.uber.org/yarpc/api/transport"
+	"go.uber.org/yarpc/encoding/raw"
+	peerbind "go.uber.org/yarpc/peer"
+	"go.uber.org/yarpc/peer/roundrobin"
+)
+
+// TransportSpec specifies how to create test clients and servers for a transport.
+type TransportSpec struct {
+	NewServerTransport func(t *testing.T, addr string) peer.Transport
+	NewClientTransport func(t *testing.T) peer.Transport
+	NewInbound         func(x peer.Transport, addr string) transport.Inbound
+	NewUnaryOutbound   func(x peer.Transport, pc peer.Chooser) transport.UnaryOutbound
+	Identify           func(addr string) peer.Identifier
+	Addr               func(x peer.Transport, ib transport.Inbound) string
+}
+
+// Test runs reusable tests with the transport spec.
```

Comment:
Why a struct instead of a struct pointer? I think you, me, and @willhug need to talk about your and Will's affinity for structs and garbage collection :p